### ISSUE TITLE
CronTriggerTimetable lost one task occasionally

### ISSUE THREAD

**Issue by potiuk** [apache/airflow#27399]
### Discussed in https://github.com/apache/airflow/discussions/27398

<div type='discussions-op-text'>

<sup>Originally posted by **AndrewTsao** October 25, 2022</sup>
### Apache Airflow version

Other Airflow 2 version (please specify below)

### What happened

My Airflow version 2.4.0. I use CronTriggerTimetable in my DAG. I find it lost one task occasionally. I inspected schedule log,

It should be run at 2022-10-24T23:10:00.000+0800. BUT not run.

```
[2022-10-24T23:08:59.661+0800] {processor.py:768} INFO - DAG(s) dict_keys(['AFlow.ajob']) retrieved from /home/bob/airflow/dags/dag.py
[2022-10-24T23:08:59.685+0800] {logging_mixin.py:117} INFO - [2022-10-24T23:08:59.684+0800] {dag.py:2573} INFO - Sync 1 DAGs
[2022-10-24T23:08:59.710+0800] {logging_mixin.py:117} INFO - [2022-10-24T23:08:59.710+0800] {dag.py:3324} INFO - Setting next_dagrun for AFlow.ajob to 2022-10-24T15:10:00+00:00, run_after=2022-10-24T15:10:00+00:00
[2022-10-24T23:08:59.734+0800] {processor.py:178} INFO - Processing /home/bob/airflow/dags/dag.py took 0.351 seconds
[2022-10-24T23:09:29.855+0800] {processor.py:156} INFO - Started process (PID=16329) to work on /home/bob/airflow/dags/dag.py
[2022-10-24T23:09:29.856+0800] {processor.py:758} INFO - Processing file /home/bob/airflow/dags/dag.py for tasks to queue
[2022-10-24T23:09:29.857+0800] {logging_mixin.py:117} INFO - [2022-10-24T23:09:29.857+0800] {dagbag.py:525} INFO - Filling up the DagBag from /home/bob/airflow/dags/dag.py
[2022-10-24T23:09:29.946+0800] {processor.py:768} INFO - DAG(s) dict_keys(['AFlow.ajob']) retrieved from /home/bob/airflow/dags/dag.py
[2022-10-24T23:09:29.969+0800] {logging_mixin.py:117} INFO - [2022-10-24T23:09:29.968+0800] {dag.py:2573} INFO - Sync 1 DAGs
[2022-10-24T23:09:29.994+0800] {logging_mixin.py:117} INFO - [2022-10-24T23:09:29.994+0800] {dag.py:3324} INFO - Setting next_dagrun for AFlow.ajob to 2022-10-24T15:10:00+00:00, run_after=2022-10-24T15:10:00+00:00
[2022-10-24T23:09:30.178+0800] {processor.py:178} INFO - Processing /home/bob/airflow/dags/dag.py took 0.327 seconds
[2022-10-24T23:10:00.368+0800] {processor.py:156} INFO - Started process (PID=16562) to work on /home/bob/airflow/dags/dag.py
[2022-10-24T23:10:00.369+0800] {processor.py:758} INFO - Processing file /home/bob/airflow/dags/dag.py for tasks to queue
[2022-10-24T23:10:00.370+0800] {logging_mixin.py:117} INFO - [2022-10-24T23:10:00.370+0800] {dagbag.py:525} INFO - Filling up the DagBag from /home/bob/airflow/dags/dag.py
[2022-10-24T23:10:00.628+0800] {processor.py:768} INFO - DAG(s) dict_keys(['AFlow.ajob']) retrieved from /home/bob/airflow/dags/dag.py
[2022-10-24T23:10:00.648+0800] {logging_mixin.py:117} INFO - [2022-10-24T23:10:00.647+0800] {dag.py:2573} INFO - Sync 1 DAGs
[2022-10-24T23:10:00.670+0800] {logging_mixin.py:117} INFO - [2022-10-24T23:10:00.670+0800] {dag.py:3324} INFO - Setting next_dagrun for AFlow.ajob to 2022-10-25T15:10:00+00:00, run_after=2022-10-25T15:10:00+00:00
[2022-10-24T23:10:00.693+0800] {processor.py:178} INFO - Processing /home/bob/airflow/dags/dag.py took 0.331 seconds
```

### What you think should happen instead

I think it lost schedule infomation when reload dag file and reschedule dag. 

### How to reproduce

1.  Change airflow.cfg `ag_dir_list_interval = 10`.
2. Add DAG as follow,

```py
from airflow import DAG
from airflow.timetables.trigger import CronTriggerTimetable
from airflow.decorators import task
import pendulum as pl

with DAG(
    dag_id = 'test-cron-2',
    # schedule_interval='10 23 * * 2-3',
    timetable=CronTriggerTimetable('* * * * *', timezone=pl.tz.get_local_timezone()),  # At 01:00 on Wednesday
    start_date=pl.today().add(days=-21),
    tags=['example'],
) as dag1:
    @task
    def test():
        print("run...")

    test()
```

![20221025132642](https://user-images.githubusercontent.com/133179/197689985-326f4736-c26f-42f7-9222-da2bbfef31c1.png)

In scheduler log, We can find on that point, scheduler is reloading dag file.

![20221025133158](https://user-images.githubusercontent.com/133179/197690750-35b7007b-0b95-4ed7-90f7-c55e74d362d2.png)

### Operating System

Centos 7.9

### Versions of Apache Airflow Providers

```
airflow-code-editor                      7.0.0
apache-airflow                           2.4.0
apache-airflow-providers-celery          3.0.0
apache-airflow-providers-common-sql      1.1.0
apache-airflow-providers-ftp             3.1.0
apache-airflow-providers-http            4.0.0
apache-airflow-providers-imap            3.0.0
apache-airflow-providers-microsoft-psrp  2.0.0
apache-airflow-providers-microsoft-winrm 3.0.0
apache-airflow-providers-mysql           3.2.0
apache-airflow-providers-redis           3.0.0
apache-airflow-providers-samba           4.0.0
apache-airflow-providers-sftp            4.0.0
apache-airflow-providers-sqlite          3.2.0
apache-airflow-providers-ssh             3.1.0
```

### Deployment

Virtualenv installation

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)
</div>

**Comment by potiuk** [comment]
Mistakenly converted to discussion. Seems legit. This is likely only happening in 1 minute interval, so not really critical but woudl be nice to find out when it happens. - thanks @andi for detailed repot. 

Could you please check @andi  if it appears in the latest version of Airflow and maybe pinpoint some other cicumstances  when it happens? How often? Do you see some usage spikes (memory/cpu etc. around that time? do you see any unsual task restarts etc.  - we need to try to get some really reproducible case in order to be able to track tha one - otherwise it is a pure quess. 

Can you try to see much more information @andi ?

**Comment by uranusjr** [comment]
Hmm, this and #27582 seems to be the same issue?

**Comment by potiuk** [comment]
I think the other issue is "first" schedule - and this one is "missing one in the middle" - it might be the same but it might be different.

**Comment by bluek1te** [comment]
Hey all, we had an issue in our configuration where our DAGs were reloading by the order of milliseconds (looking at last_parsed_time in the database) and we observed this issue happening a lot. The only DAG that failed to run in this circumstance were the ones that used CronTriggerTimetable. (We were using a test that should run every minute). I believe that if you set the DAG reload time very, very low, you will see this problem occur much more frequently.

**Comment by johnwarburton** [comment]
Can confirm when trying to schedule a DAG exactly one with `CronTriggerTimetable` airflow [v2.4.2] in **standalone mode** (where `dag_dir_list_interval = 300` seems to be ignored)
```
import subprocess
from airflow import DAG
from airflow.decorators import dag, task
from airflow.timetables.trigger import CronTriggerTimetable
from datetime import datetime

@task
def mytask():
  bash_command=f'/bin/echo run once'
  EXEC_OUT=(subprocess.check_output(bash_command, shell=True)).decode('utf-8')
  
@dag(
  dag_id=f'run_once',
  start_date=datetime(2022,1,1),
  timetable=CronTriggerTimetable('03 18 16 11 *', timezone='Australia/Sydney'),
  catchup=False,
)
def run_once_dag():
  mytask()
  
run_once_dag()  
```
In the minute before 18:03, the DAG was processed 22 times

The first processing at 18:03, the `next_dagrun` was set to the same time one year later, without executing the DAG
```
[2022-11-16T18:02:50.448+1100] {processor.py:154} INFO - Started process (PID=11018) to work on /export/home/jwarburt/airflow-venv/dags/run-once.py
[2022-11-16T18:02:50.506+1100] {logging_mixin.py:120} INFO - [2022-11-16T18:02:50.506+1100] {dag.py:3336} INFO - Setting next_dagrun for run_once to 2022-11-16T07:03:00+00:00, run_after=2022-11-16T07:03:00+00:00
[2022-11-16T18:02:50.609+1100] {processor.py:176} INFO - Processing /export/home/jwarburt/airflow-venv/dags/run-once.py took 0.164 seconds
[2022-11-16T18:02:53.836+1100] {processor.py:154} INFO - Started process (PID=11026) to work on /export/home/jwarburt/airflow-venv/dags/run-once.py
[2022-11-16T18:02:53.894+1100] {logging_mixin.py:120} INFO - [2022-11-16T18:02:53.894+1100] {dag.py:3336} INFO - Setting next_dagrun for run_once to 2022-11-16T07:03:00+00:00, run_after=2022-11-16T07:03:00+00:00
[2022-11-16T18:02:54.004+1100] {processor.py:176} INFO - Processing /export/home/jwarburt/airflow-venv/dags/run-once.py took 0.171 seconds
[2022-11-16T18:02:59.105+1100] {processor.py:154} INFO - Started process (PID=11032) to work on /export/home/jwarburt/airflow-venv/dags/run-once.py
[2022-11-16T18:02:59.163+1100] {logging_mixin.py:120} INFO - [2022-11-16T18:02:59.163+1100] {dag.py:3336} INFO - Setting next_dagrun for run_once to 2022-11-16T07:03:00+00:00, run_after=2022-11-16T07:03:00+00:00
[2022-11-16T18:02:59.272+1100] {processor.py:176} INFO - Processing /export/home/jwarburt/airflow-venv/dags/run-once.py took 0.171 seconds
[2022-11-16T18:03:00.699+1100] {processor.py:154} INFO - Started process (PID=11034) to work on /export/home/jwarburt/airflow-venv/dags/run-once.py
[2022-11-16T18:03:00.845+1100] {logging_mixin.py:120} INFO - [2022-11-16T18:03:00.845+1100] {dag.py:3336} INFO - Setting next_dagrun for run_once to 2023-11-16T07:03:00+00:00, run_after=2023-11-16T07:03:00+00:00
[2022-11-16T18:03:00.865+1100] {processor.py:176} INFO - Processing /export/home/jwarburt/airflow-venv/dags/run-once.py took 0.169 seconds
```

If I could slow down the processing interval, that might help?

**Comment by RNHTTR** [comment]
@johnwarburton Have you had any luck with this? Seeing the same issue in 2.4.2 after switching from a plain cron expression (i.e. `schedule_interval=0 15 * * *`) to using CronTriggerTimetable directly with a timezone (i.e. `timetable=CronTriggerTimetable("0 15 * * *", timezone="America/Los_Angeles")`). I noticed that all of the examples in this thread have a timezone attached.

**Comment by johnwarburton** [comment]
Hi @RNHTTR - no, I am waiting to see if there is a fix in the milestones

`CronTriggerTimetable` must have a timezone attached:
```
   |     timetable=CronTriggerTimetable('45 4 2 12 *'),
   | TypeError: __init__() missing 1 required keyword-only argument: 'timezone'
```

**Comment by RNHTTR** [comment]
Apologies -- I thought that using a cron expression such as `schedule_interval=0 15 * * *` uses `CronTriggerTimetable` under the hood, but it looks like it actually users [CronDataIntervalTimetable](https://github.com/apache/airflow/blob/main/airflow/timetables/interval.py#L112)

**Comment by fatmumuhomer** [comment]
I was chatting with Jarek earlier today on Slack about a similar issue I saw over the weekend.

`schedule=CronTriggerTimetable('15 12 * * *', timezone=pendulum.timezone('America/New York')) `

In my scenario I had deployed the DAG late on Friday and manually triggered it since the next run wasn't scheduled until Saturday. It then successfully scheduled on Saturday but not on Sunday or today (Monday). I am going through logs and will see if it runs on Tuesday.

**Comment by RNHTTR** [comment]
One commonality is all timezones are non-utc. Not sure if that's indicative of something, but I think it's worth noting.

**Comment by scodesido-at-proton** [comment]
We have observed the same issue with a UTC timezone, so I don't think it is the issue, e.g.
```
timetable=CronTriggerTimetable("*/15 * * * *", timezone="UTC")
```

Are there any tips on how to deal with this reliably, or should we just expect `CronTriggerTimetable` to silently fail every now and then? In that case, maybe it would be good to update the docs [here](https://airflow.apache.org/docs/apache-airflow/stable/concepts/timetable.html#crontriggertimetable) to warn about the inconsistent behaviour?

**Comment by potiuk** [comment]
Also observec in #28293

**Comment by isaac-florence** [comment]
@potiuk do you have any advice on how to proceed with managing this bug?

**Comment by potiuk** [comment]
No. Not until it gets investigated and fixed.

**Comment by Gollum999** [comment]
For what it's worth, prior to 2.4 I wrote my own timetable that acts almost exactly like `CronTriggerTimetable`, and I experienced a similar issue, so I'll mention my findings here.

In my case the problem was that there was a race condition between the DagProcessor and the Scheduler when calling `Timetable.next_dagrun_info`. Depending on when each thread evaluated the timetable, the DagProcessor could end up "bumping" the DAG's `next_dagrun_create_after` timestamp in the DB before the Scheduler had a chance to schedule the current interval.  Then when the Scheduler would do it's query to determine which DagRuns to schedule, it wouldn't find the run that got "skipped".  In my case, I believe the bug was a call to `DateTime.utcnow()` that I had to align to the previous interval boundary.

Also worth noting that my bug only happened with `catchup=False`, but that could have just been a quirk with my implementation.

**Comment by bluek1te** [comment]
Hello friends. Thinking about @Gollum999's explanation of the problem case, I ended up being able to fix the missed run issue by doing this:

```diff
diff --git a/airflow/timetables/trigger.py b/airflow/timetables/trigger.py
index 86ab58cc4d..67dd239f89 100644
--- a/airflow/timetables/trigger.py
+++ b/airflow/timetables/trigger.py
@@ -93,7 +93,8 @@ class CronTriggerTimetable(CronMixin, Timetable):
             if restriction.earliest is not None and current_time < restriction.earliest:
                 next_start_time = self._align_to_next(restriction.earliest)
             else:
-                next_start_time = self._align_to_next(current_time)
+                prev_start_time = self._align_to_prev(current_time)
+                next_start_time = self._align_to_next(prev_start_time)
         if restriction.latest is not None and restriction.latest < next_start_time:
             return None
         return DagRunInfo.interval(next_start_time - self._interval, next_start_time)
```
Instead of aligning next with the current time and running into the race condition with the scheduler, it might be better if we align next to the previously scheduled time.
I would like to open this up to a formal PR if y'all think this is sufficient. Thoughts?

**Comment by potiuk** [comment]
cc: @uranusjr ? WDYT this looks plausibke I guess ^^

**Comment by LMnet** [comment]
Is there any workaround to fix this issue? I'm experiencing the same problem with my daily scheduled DAGs: sometimes they randomly don't run.

I have a `catchup=False` and I can't switch to `CronDataIntervalTimetable`.

**Comment by mai-nakagawa** [comment]
My gut feeling is also that the root cause is somewhere in DagProcessor and/or Scheduler.

Anyway this problem doesn't happen in the `CronDataIntervalTimetable`. I think the reason is because `CronDataIntervalTimetable` runs a last DAG run if it missed even if `catchup=False`.

Maybe we need to implement the similar functionality in `CronTriggerTimetable`, like a `anacron`. What do you think? @uranusjr @potiuk

**Comment by potiuk** [comment]
Hard to say.

**Comment by LMnet** [comment]
May it be somehow related to the fact that `CronTriggerTimetable` by default has an empty date interval? As far as I understand most (if not all) built-in timetables have non-empty date intervals.

To test this hypothesis I set up small (10 minutes) date intervals for my daily jobs. Will report how it goes when I have some results.

**Comment by uranusjr** [comment]
I don’t _think_ the interval is relevant; the scheduler never looks at it, but only pass it directly into a run’s context.

**Comment by uranusjr** [comment]
This is the logic to calculate the next run for `catchup=False`, for reference:

```python
start_time_candidates = [self._align_to_next(DateTime.utcnow())]
if last_automated_data_interval is not None:
    start_time_candidates.append(self._get_next(last_automated_data_interval.end))
if restriction.earliest is not None:
    start_time_candidates.append(self._align_to_next(restriction.earliest))
next_start_time = max(start_time_candidates)
``

The two `if` blocks should be fine since they match the logic for `catchup=True`. So I wonder if the problem is from the `max` call—maybe under some edge cases that `max` would wrongly select the `next(utcnow)` result (first line) instead of the actual schedule it should use—when the scheduler took too much time to schedule the previous run, perhaps?

Also worth noting that `CronDataIntervalTimetable` (the old one) uses slightly different logic to account for the current time, and that timetable seems to be bug-free.

**Comment by LMnet** [comment]
> May it be somehow related to the fact that `CronTriggerTimetable` by default has an empty date interval? As far as I understand most (if not all) built-in timetables have non-empty date intervals.
> 
> To test this hypothesis I set up small (10 minutes) date intervals for my daily jobs. Will report how it goes when I have some results.

FYI: even with the non-zero date interval I still faced the issue. So, this is not the root cause of this problem.

Overall it looks like the only real workaround for now is to not use `CronTriggerTimetable` at all.

But there is a bigger problem with this bug in my opinion. Since we are not 100% sure what is the root cause, all other timetables may be affected by this bug too. And since the main purpose of Airflow is to reliably schedule jobs, I think this is a major problem: with this bug we can't say that Airflow guarantees that user's jobs will be scheduled as requested. They can be randomly skipped. And even if only `CronTriggerTimetable` is affected, users can create custom timetables, and they could be affected by this bug too.

**Comment by uranusjr** [comment]
Contributions are always welcomed.

**Comment by uranusjr** [comment]
Forgot to mentioned the one-line change. This line

https://github.com/apache/airflow/blob/4a44343731144a7a7dc7fff7e3ed01663d4dd2e1/airflow/timetables/trigger.py#L92

should be changed to use `_aligned_to_prev` instead.

**Comment by uranusjr** [comment]
Yeah. Or maybe we can addd a flag for this? Not sure what it should be called though.

**Comment by shmcsensei** [comment]
I've been testing the following change to fix this issue seems to be working. Its similar to what has been suggested:

```python
def next_dagrun_info(
    self,
    *,
    last_automated_data_interval: Optional[DataInterval],
    restriction: TimeRestriction,
) -> Optional[DagRunInfo]:
    if restriction.catchup:
        if last_automated_data_interval is None:
            if restriction.earliest is None:
                return None
            next_start_time = self._align_to_next(restriction.earliest)
        else:
            next_start_time = self._get_next(last_automated_data_interval.end)
    else:
        current_time = DateTime.utcnow()
        if restriction.earliest is not None and current_time < restriction.earliest:
            next_start_time = self._align_to_next(restriction.earliest)
        else:
            if last_automated_data_interval is None:
                # The DAG has never run before.
                next_start_time = self._align_to_next(current_time)
            elif last_automated_data_interval.end == self._align_to_prev(current_time):
                # The last run is the same as the previously scheduled run. Everything is working fine.
                next_start_time = self._align_to_next(current_time)
            else:
                # The last run is NOT the same as the previously scheduled run, we must have skipped it.
                next_start_time = self._align_to_prev(current_time)
                LOGGER.warn('DAG skipped previous schedule, rescheduling. last_automated_data_interval.end:"%s" next_start_time:"%s"', last_automated_data_interval.end, next_start_time)

    if restriction.latest is not None and restriction.latest < next_start_time:
        return None
    return DagRunInfo.exact(next_start_time)

```

Specifically this addition:

```python
 if last_automated_data_interval is None:
    # The DAG has never run before.
    next_start_time = self._align_to_next(current_time)
elif last_automated_data_interval.end == self._align_to_prev(current_time):
    # The last run is the same as the previously scheduled run. Everything is working fine.
    next_start_time = self._align_to_next(current_time)
else:
    # The last run is NOT the same as the previously scheduled run, we must have skipped it.
    next_start_time = self._align_to_prev(current_time)
    LOGGER.warn('DAG skipped previous schedule, rescheduling. last_automated_data_interval.end:"%s" next_start_time:"%s"', last_automated_data_interval.end, next_start_time)
```

I added the logging so I could monitor it and alert if it happens, just in case. Not sure if this could be done better though, I'm open to suggestions.

**Comment by uranusjr** [comment]
```python
elif last_automated_data_interval.end == self._align_to_prev(current_time):
    # The last run is the same as the previously scheduled run. Everything is working fine.
    next_start_time = self._align_to_next(current_time)
```

This part has a bug when `current_time` is _exactly_ `last_automated_data_interval.end`, `next_start_time` would be set to `current_time` again (because the align functions will try to not move the time if it already lands on the cron). Very unlikely but still. I think it can be avoided with `self._get_next(last_automated_data_interval.end)` instead, to ensure the logic progresses one interval ahead.

The `else` block is essentially the same as the change discused above and looks good to me.

**Comment by shmcsensei** [comment]
Thanks @uranusjr for the feedback! I'll look into it, good points.

**Comment by BohdanSemonov** [comment]
The bug occurs due to the fact that seconds and microseconds are not reset here: `start_time_candidates = [self._align_to_next(DateTime.utcnow())]` and a condition `if self._get_prev(next_time) != current:` in the method `_align_to_next` chooses `next_time` instead `current`.

The bug fix is:
`start_time_candidates = [self._align_to_next(DateTime.utcnow().replace(second=0, microsecond=0))]`
I've already tested the fix - it works properly.

**Comment by snippins** [comment]
@BohdanSemonov I tried to applied your fix and do the following test.

+ current time is 10:00
+ switch the cron expression to '2 10 * * *'
+ wait for scheduler to update (done around 10:01)
+ pause the dag
+ wait until 10:03
+ resume the dag

The dag start to run, maybe I misunderstand something but I thought the entire point of CronTriggerTimetable is to prevent this?

**Comment by BohdanSemonov** [comment]
@snippins
The situation you describe is the Airflow feature when `catchup=True`.
Try with `catchup=False`.

**Comment by Tindarid** [comment]
Hey @BohdanSemonov , if the fix works, could you please submit a PR? I think this issue is quite critical, and I also observe such behavior from CronTriggerTimetable

**Comment by snippins** [comment]
> The situation you describe is the Airflow feature when `catchup=True`.

I have catchup=False already.

**Comment by BohdanSemonov** [comment]
> > The situation you describe is the Airflow feature when `catchup=True`.
> 
> I have catchup=False already.

Sorry, maybe we didn't understand each other, but I observe such behavior when `catchup=True`, otherwise I could not get the described DAG run.

**Comment by BohdanSemonov** [comment]
> Hey @BohdanSemonov , if the fix works, could you please submit a PR? I think this issue is quite critical, and I also observe such behavior from CronTriggerTimetable

Done.

**Comment by RNHTTR** [comment]
I'm closing this as it appears to have been fixed in 2.7.1 by #33404 but there was a typo in the description that allowed this Issue to stay open.
